File: lib/digest_mailer.rb
# frozen_string_literal: true

require 'aws-sdk-ses'

class DigestMailer
  SES_RECIPIENT_LIMIT = 50
  private_constant :SES_RECIPIENT_LIMIT

  FROM = 'hndigest@samshadwell.com'
  private_constant :FROM

  REPLY_TO = 'hi@samshadwell.com'
  private_constant :REPLY_TO

  ENCODING = 'UTF-8'
  private_constant :ENCODING

  def initialize(ses_client:)
    @ses_client = ses_client
  end

  def send_mail(renderer:, recipients:)
    recipients.each_slice(SES_RECIPIENT_LIMIT) do |recipients_slice|
      puts 'Sending mail via SES...'
      response = ses_client.send_email({
        source: FROM,
        destination: {
          bcc_addresses: recipients_slice
        },
        reply_to_addresses: [REPLY_TO],
        return_path: REPLY_TO,
        message: {
          subject: {
            data: renderer.subject,
            charset: ENCODING,
          },
          body: {
            html: {
              data: renderer.content,
              charset: ENCODING,
            }
          }
        }
      })
      puts "Success! message_id=#{response.message_id}"
    end
  end
end

File: src/digest_mailer.rs
use aws_sdk_sesv2::Client;
use aws_sdk_sesv2::types::{Body, Content, Destination, EmailContent, Message, MessageHeader};

use anyhow::{Context, Result};
use tracing::info;

pub struct DigestMailer {
    ses_client: Client,
    from_address: String,
    reply_to_address: String,
}

impl DigestMailer {
    pub fn new(ses_client: Client, from_address: String, reply_to_address: String) -> Self {
        Self {
            ses_client,
            from_address,
            reply_to_address,
        }
    }

    /// Send an email to a single recipient with RFC 8058 unsubscribe headers.
    pub async fn send_mail(
        &self,
        subject: &str,
        html_content: &str,
        text_content: &str,
        recipient: &str,
        unsubscribe_url: &str,
    ) -> Result<()> {
        // Build RFC 8058 List-Unsubscribe headers
        let list_unsubscribe_header = MessageHeader::builder()
            .name("List-Unsubscribe")
            .value(format!("<{}>", unsubscribe_url))
            .build()?;

        let list_unsubscribe_post_header = MessageHeader::builder()
            .name("List-Unsubscribe-Post")
            .value("List-Unsubscribe=One-Click")
            .build()?;

        // Build email content
        let subject_content = Content::builder().data(subject).charset("UTF-8").build()?;
        let html_body = Content::builder()
            .data(html_content)
            .charset("UTF-8")
            .build()?;
        let text_body = Content::builder()
            .data(text_content)
            .charset("UTF-8")
            .build()?;

        let body = Body::builder().html(html_body).text(text_body).build();
        let message = Message::builder()
            .subject(subject_content)
            .body(body)
            .headers(list_unsubscribe_header)
            .headers(list_unsubscribe_post_header)
            .build();

        let destination = Destination::builder().to_addresses(recipient).build();

        let email_content = EmailContent::builder().simple(message).build();

        let response = self
            .ses_client
            .send_email()
            .from_email_address(&self.from_address)
            .reply_to_addresses(&self.reply_to_address)
            .destination(destination)
            .content(email_content)
            .send()
            .await
            .context(format!("Failed to send email to {}", recipient))?;

        info!(
            message_id = ?response.message_id(),
            recipient = recipient,
            "Email sent"
        );

        Ok(())
    }
}
